nselib/smb2.lua

```diff
@@ -0,0 +1,410 @@
+    (overrides['SessionId'] or smb['SessionId'] or 0),  -- 8 bytes: SessionId.
+    (overrides['Signature'] or '1234567890123456')     -- 16 bytes: Signature.
+    )
+
+  return header
+end
+
+---
+-- Sends a SMB2 packet 
+-- @param smb        The SMB object associated with the connection
+-- @param header     The header encoded with <code>smb_encode_sync_header</code>.
+-- @param data       The data.
+-- @param overrides  Overrides table.
+-- @return (result, err) If result is false, err is the error message. Otherwise, err is
+--        undefined
+---
+function smb2_send(smb, header, data, overrides)
+  overrides = overrides or {}
+  local body               = header .. data
+  local attempts           = 5
+  local status, err
+
+  local out = string.pack(">I<c" .. #body, #body, body)
+  repeat
+    attempts = attempts - 1
+    stdnse.debug3("SMB: Sending SMB packet (len: %d, attempts remaining: %d)", #out, attempts)
+    status, err = smb['socket']:send(out)
+  until(status or (attempts == 0))
+
+  if(attempts == 0) then
+    stdnse.debug1("SMB: Sending packet failed after 5 tries! Giving up.")
+  end
+
+  return status, err
+end
+
+---
+-- Reads the next SMB2 packet from the socket, and parses it into the header and data.
+-- Netbios handling based taken from smb.lua.
+--
+-- @param smb The SMB object associated with the connection
+-- @param read_data [optional] Return data section. Set to false if you only need the header. Default: true
+-- @return (status, header, data) If status is true, the header,
+--         and data are all the raw arrays of bytes.
+--         If status is false, header contains an error message and data is undefined.
+---
+function smb2_read(smb, read_data)
+  local status
+  local pos, netbios_data, netbios_length, length, header, parameter_length, parameters, data_length, data
+  local attempts = 5
+
+  stdnse.debug3("SMB2: Receiving SMB2 packet")
+
+  -- Receive the response -- we make sure to receive at least 4 bytes, the length of the NetBIOS length
+  smb['socket']:set_timeout(TIMEOUT)
+
+  -- perform 5 attempt to read the Netbios header
+  local netbios
+  repeat
+    attempts = attempts - 1
+    status, netbios_data = smb['socket']:receive_buf(match.numbytes(4), true);
+
+    if ( not(status) and netbios_data == "EOF" ) then
+      return false, "SMB2: ERROR: Server disconnected the connection"
+    end
+  until(status or (attempts == 0))
+
+  -- Make sure the connection is still alive
+  if(status ~= true) then
+    return false, "SMB2: Failed to receive bytes after 5 attempts: " .. netbios_data
+  end
+
+  -- The length of the packet is 4 bytes of big endian (for our purposes).
+  -- The NetBIOS header is 24 bits, big endian
+  netbios_length, pos   = string.unpack(">I", netbios_data)
+  if(netbios_length == nil) then
+    return false, "SMB2: ERROR:Server returned less data than it was supposed to"
+  end
+  -- Make the length 24 bits
+  netbios_length = bit.band(netbios_length, 0x00FFFFFF)
```

Comment:
Please use [native Lua 5.3 bit operations](https://www.lua.org/manual/5.3/manual.html#3.4.2). The bit library is deprecated.